tools/node-modules

```diff
@@ -0,0 +1,182 @@
+#!/bin/sh
+
+set -eu
+
+GITHUB_BASE="${GITHUB_BASE:-cockpit-project/cockpit}"
+CACHE_DIR="${XDG_CACHE_HOME-${HOME}/.cache}/cockpit-dev/node-cache.git"
+CLONE_URL="https://github.com/${GITHUB_BASE%/*}/node-cache"
+
+cmd_remove() {
+    # if we did this for ourselves the rm is enough, but it might be the case
+    # that someone actually used git-submodule to fetch this, so clean up after
+    # that as well.  NB: deinit nicely recreates the empty directory for us.
+    rm -rf node_modules
+    git submodule deinit node_modules
+    rm -rf "$(git rev-parse --absolute-git-dir)/modules/node_modules"
+}
+
+cmd_checkout() {
+    # we check out the node_modules corresponding to the gitlink in the index
+    local sha="$(git submodule status --cached node_modules | cut -c2-41)"
+    local tag="sha-${sha}"
+
+    if [ ! -d "${CACHE_DIR}" ]; then
+        [ "${V-}" != 0 ] || echo "  INIT     ${CACHE_DIR}"
+        mkdir -p "${CACHE_DIR}"
+        git --git-dir "${CACHE_DIR}" init ${quiet}
+        git --git-dir "${CACHE_DIR}" remote add origin "${CLONE_URL}"
+    fi
+
+    # fetch the tag if we don't already have it
+    if ! git --git-dir "${CACHE_DIR}" rev-parse --verify --quiet "${tag}" >/dev/null; then
+        [ "${V-}" != 0 ] || echo "  FETCH    node_modules  [sha: ${sha}]"
+        # fetch by sha to prevent us from downloading something we don't want
+        # the tag should come automatically with it
+        git --git-dir "${CACHE_DIR}" fetch ${quiet} --tags origin "${sha}"
+    fi
+
+    # double check that the tag is what it claims to be
+    test "$(git --git-dir "${CACHE_DIR}" rev-parse "${tag}")" = "${sha}"
+
+    # verify that our package.json is equal to the one the cached node_modules
+    # was created with, unless --force is given
+    if [ "${1-}" != "--force" ]; then
+        local our_package_json_sha="$(cat package.json | sha1sum)"
+        local their_package_json_sha="$(git --git-dir "${CACHE_DIR}" cat-file -p "${sha}":.package.json | sha1sum)"
```

Comment:
Can we use sha256sum here? We still get the [occasional audit](https://bugzilla.redhat.com/show_bug.cgi?id=1934937) for the long-broken SHA-1, so let's please avoid that noise in `git grep`.

Reply:
Weird.  At some point I thought I had changed that, because ya, we have no need to match git here.  Will do.